### DISCUSSION TITLE
[Feature] Deletion of Detected Faces

### DISCUSSION THREAD

**Discussion by AllSparkII** [immich-app/immich#13868] [Feature Request]
### I have searched the existing feature requests to make sure this is not a duplicate request.

- [X] Yes

### The feature

**TLDR: I am looking for a way to (permanently) remove people from the "Peoples" Tab, without regrouping them or toggling their visibility. Reason being: Strangers fill my "invisible" category to a point where it's useless**

I would be interested in a feature that (permanently) removes people from my "People"-Tab. I tuned my AI-parameters to a point, where over 90% of faces get recognized in my library of 42.000 Images. This results in a problem: Over 6k faces are shown in the Tab. I have no interest in many of them, as they are f.e. bystanders at some sports event or just random people strolling through cities in my vacation pictures. Setting higher recognition boundaries or using lower accuracy models decreases these kinds of results, but unfortunately the people I am interested in get recognized worse too. For this, I'd like to remove people without grouping them.

There is a feature to toggle the visibility of certain people. I use it for people I usually don't search for or for people I don't want to pop up randomly (f.e. a deceased loved one). Moving ~5000 unknown people in there would be bad for finding what I am actually looking for and bad for conciseness. Clustering all unknown people into a single Person would result in new pictures to get randomly grouped in there.

### Platform

- [X] Server
- [X] Web
- [X] Mobile

**Comment by mertalev** [comment]
If it's that bad, you might want to reset facial recognition with more normal settings. What did you change the settings to?

**Reply by AllSparkII** [reply]
Thank you for your reply! My settings are the following: Buffalo L with 0.5 Minimum detect and 0.5 maximum distance. It needs 2 Faces for a person to be created.

I acknowledge that these settings are chosen to find the highest amount of faces possible - but that's what facial recognition is made for, isn't it? Using this technique, I got an increase of about 20 % in detected pictures of myself and some people close to me (compared to G Fotos and Synology Photos). It's not "bad" in a sense of false positives (1 per 3000 created people)! Upping the minimal amout of detected faces sadly won't work for me, as many of my photos are Bursts of Scenes in Sports.

It´s more of a manageability problem. I wouldn't mind going through it once and remove the faces I won´t ever look for.

**Reply by mertalev** [reply]
The "min recognized faces" setting is your best friend here, but burst photos are a challenge. My recommendation is to set this to 15 and reset facial recognition. This will avoid these people from being created and moreover prevent the blobs of random people you mention. Depending on how many burst photos there are for the same scene, you might need to tweak this number a bit.

**Reply by mertalev** [reply]
Also general tip that you can set this to a high number and later lower it if need be without resetting, but you can't do the opposite without resetting if you want to apply it to the current clustering. So feel free to use a high number and work from there.

**Reply by AllSparkII** [reply]
Thank you for tip!

Using the higher "min recognized faces" would result in losing People I'd like to know discovered and still having the burst artifacts. There are two workarounds, one is already implemented:

1. Set the "min recognized faces" high (like 15) and go through every picture in the gallery to manually to retag people lost due to <15 pictures and manually merge Pictures that have faces in them but weren't correctly grouped. These people would disappear in the People-Menu due to the high "min recognized faces". 

2. (not implemented, but would be a great feature i.m.o.) Having the "min recognized faces" low (like my 2) and initially use the People-Menu as overview over all recognized faces. Tag the people you want to be displayed, merge if people weren't grouped correctly. Many faces won´t be of any value, but could be removed out of the list with a simple click.

The outcome of both versions should be the same, or at least pretty close. But Version two would only require looking at 6000 "faces", Version 1 would require looking at all 40 k Pictures.
I am all about having the most faces detected and grouped correctly. My parameters were specifically tuned for this purpose. I know and understand that my approach isn't for everybody, and many people would rather have 20 faces not detected than doing so much work. I still believe, that this would be a valuable feature for some people.

**Reply by mertalev** [reply]
Oh, I agree that being able to delete people would be helpful. But if it got to the point where you have thousands of irrelevant people, it would be a fairly significant amount of work to sort through this by deleting them.

Since you mentioned that you tried a higher value for this setting and it affected relevant people, here's another tip from @aviv926: if you set this to a high value initially and reset facial recognition, then lower it and run "missing" for facial recognition, the result is actually better than using the lower value to start. This is likely because the initial clustering is higher quality and the second pass has a better foundation to work with. It might let you increase your current value of 2 to something a bit higher.

**Reply by multi-suggester** [reply]
@AllSparkII just a suggestion but if you dont have too many named person , i might suggest making a tag structure and having it as a backup and then reset the people recognition (which is what i did while testing different recognition values) , the main structure could be something like /people /-relatives --me --family   /-friends /-job sort of thing would be probably sufficient

**Comment by CyberCreator** [comment]
This would be a useful option, I support it.

I would like to make a suggestion. You can not delete strangers, but place them in the appropriate folder with "unnecessary", so that in the future, if suddenly needed, you can get the right face, or if the "stranger" suddenly becomes "familiar", then you will not need to re-render all the photos where his face was deleted. Then it will be enough to upload a photo with this person, and then select his face. (This action should be associated with the previously hidden face in other photos)

**Reply by ffchung** [reply]
No, you will never need them. When you go to travel or take photo with family, you will have main people you need. but other strangers you will never need them. Like you take photo on the street, more people around. I just want remove them and I dont need review it everytime. And if the case I need them again, I can just rescan the face.

**Reply by ffchung** [reply]
Also if I delete the person I don't need. It will decrease the chance to put the face to wrong person. I just start to try immich, I put 1/10 of my photo. and there have 5k+ group of person and only around 100 person I need. and some of them with wrong face in person, I need check them one by one.

**Reply by ffchung** [reply]
All though I want to delete.

Face on the Board.
![Screenshot 2024-12-02 at 10 13 25 PM](https://github.com/user-attachments/assets/b7f37a9b-e3ff-45ba-b1fb-87ab4ec9c9b7)

Doll Face
![Screenshot 2024-12-02 at 10 13 30 PM](https://github.com/user-attachments/assets/66d0aeea-87ef-4c74-9866-03d186955ec8)

Stranger  mix with my other face.
![Screenshot 2024-12-02 at 10 20 46 PM](https://github.com/user-attachments/assets/9b18ad9c-863b-4160-aeaf-642545e77ee8)

Stranger
![Screenshot 2024-12-02 at 10 13 38 PM](https://github.com/user-attachments/assets/5848ed39-12b2-4372-99b5-6cfcd8964d79)

**Comment by ffchung** [comment]
at this time, I am direct delete record in Database.

I grouped all face I want delete to one person. and delete the person in database. (It will delete all face related person automatically)

`delete from person where "id" = '7194cc37-3140-4c58-ad9e-90a0ffd13028';`

**Reply by ffchung** [reply]
Also need delete record in asset_faces .

```
delete from asset_faces where "personId" = '7194cc37-3140-4c58-ad9e-90a0ffd13028';
delete from person where "id" = '7194cc37-3140-4c58-ad9e-90a0ffd13028';
```

**Reply by Ole-png** [reply]
How do you get the ID of the person? I can't seem to find it anywhere. This does seem useful as an intermediate solution until it is implemented.

**Reply by ffchung** [reply]
@Ole-png  
Go to People Page, you can see the id at url link.
/people/7194cc37-3140-4c58-ad9e-90a0ffd13028?previousRoute=/people

**Reply by Ole-png** [reply]
Thanks, 
for anyone trying this start with: `docker exec -it immich_postgres psql -U postgres -d immich`
and if your lasy just send all the peopel links to chat gpt and let him gather and insert all the Ids worked great for me.

**Reply by ffchung** [reply]
@Ole-png How can I use with chatgpt ?

**Reply by Ole-png** [reply]
@ffchung like that: 
Hey Chat GPT, please do the following:
Extract the UUID from each link (the part right after /people/ up to a ? or the end of the string).
Generate two SQL commands for each UUID:
One to delete from asset_faces:
DELETE FROM asset_faces WHERE "personId" = '<UUID>';  
One to delete from person:
DELETE FROM person WHERE "id" = '<UUID>';  
Here are the links:
https://immich.XXX.XXX/people/fbbf18bd-0e06-4f67-b010-3b28ae21adb3?previousRoute=/people
https://immich.XXX.XXX/people/93215f48-1d72-46ab-a7f1-f46c8d5b9b16?previousRoute=/people 
https://immich.XXX.XXX/people/9b66c6b5-bab1-4727-8940-12ca064c4069?previousRoute=/people
....

**Reply by ffchung** [reply]
@Ole-png Thanks for share. 
Actually I make a php page for I check the face and delete.

// 1. Select Person without name
$sql = "select * from person where \"isHidden\" = $isHidden and \"ownerId\"::text='$owner' and \"name\" = '' order by \"updatedAt\" desc limit 10 offset $offset";

// 2. Show Person Page to check 
https://immich.xxx.xxx/people/$personId?previousRoute=/people

// 3. php page for delete

<img width="1616" alt="Screenshot 2025-02-18 at 3 19 23 PM" src="https://github.com/user-attachments/assets/93b89045-f1b9-4442-b75a-67a49389812e" />

**Reply by ffchung** [reply]
> Extract the UUID from each link (the part right after /people/ up to a ? or the end of the string).
> Generate two SQL commands for each UUID:
> One to delete from asset_faces:
> DELETE FROM asset_faces WHERE "personId" = '';
> One to delete from person:
> DELETE FROM person WHERE "id" = '';
> Here are the links:
> https://immich.XXX.XXX/people/fbbf18bd-0e06-4f67-b010-3b28ae21adb3?previousRoute=/people
> https://immich.XXX.XXX/people/93215f48-1d72-46ab-a7f1-f46c8d5b9b16?previousRoute=/people
> https://immich.XXX.XXX/people/9b66c6b5-bab1-4727-8940-12ca064c4069?previousRoute=/people

I just try your method, and it working well. I never try it this way. 👍
<img width="719" alt="Screenshot 2025-02-19 at 2 40 36 PM" src="https://github.com/user-attachments/assets/4ff2b7c4-4f4e-4127-b32e-dffea226e5ba" />

**Reply by Ole-png** [reply]
@ffchung Actually, that's not too smart, tell it to put everything in one box so you only have to copy and paste once. It really helps when deleting hundreds of people.  

Could you explain a bit more about your PHP tactic? I'm not really getting it. Where do you run the command? It says "unknown command," and on which page do you access it?  

Thanks!

**Reply by ffchung** [reply]
it not command, it is php webpage. 

index.php
```
<?
ini_set('display_errors', '1');
ini_set('display_startup_errors', '1');
error_reporting(E_ALL);

$immBase = "http://192.168.1.1:48080";


// immich ownerId 
$owner = “OWNER_ID;
$isHidden = "false";

$host = "192.168.1.1”;
$port = 45432;
$user = "postgres";
$pwd = “password”;
$db = "immich";

$dsn = "pgsql:host=$host;port=$port;dbname=$db;user=$user;password=$pwd";


$offset = 0;

if(isset( $_REQUEST['offset']))
{
    $in_os = trim($_REQUEST['offset']);
    if(is_numeric($in_os))
    {
        $offset = (int) $in_os;
        if($offset < 0)
        {
                $offset = 0;
        }
    }
}

echo "<html><script>

window.addEventListener(\"keydown\", logKey);

function logKey(e)
{
    switch (e.keyCode) {
	case 38:
	case 116:
		window.location.reload();
		return;
	case 39:
		window.location.href = '?offset=".($offset+10)."';
                break;
        case 37:
		window.location.href = '?offset=".($offset-10)."';
                break;
    }
}  

</script>"; 
echo "<body>";   
try{
   
    $conn = new PDO($dsn);
     
    if($conn){
        $result = $conn->query("SELECT version()");
        $version = $result->fetchColumn();
        echo "<div>";
        echo "Success Connected to <strong>$db</strong>！<br>";
 	echo "</div>";


	
        $sql = "select * from person where \"isHidden\" = $isHidden and \"ownerId\"::text='$owner' and \"name\" = '' order by \"updatedAt\" desc limit 10 offset $offset";

	echo "<table>"; 

	foreach ($conn->query($sql) as $row) {
        $id = $row['id'];

        print "<tr>"; 
            print "<td><img src=\"$immBase/api/people/$id/thumbnail\" width=40 height=40 ></td>";
            print "<td>".$id . "</td>";
            print "<td>".$row['name'] . "</td>";
            print "<td><a target=\"pview\" href=\"$immBase/people/$id?previousRoute=/people\">View</a></td>";
            print "<td><a target=\"presult\" href=\"deleteperson.php?id=$id\">Delete</a></td>";
        print "</tr>"; 

        }
        echo "</table>"; 

    }
}catch (PDOException $e){
    $errorMessage = $e->getMessage();
    echo $errorMessage;
}
echo "<br><br><div>$offset - <a href=\"?offset=".($offset+10)."\">Next</a> - <a href=\"?offset=".($offset+100)."\">Next100</a> - <a href=\"?offset=".($offset+1000)."\">Next1000</a></div><br><button onClick=\"window.location.reload()\">Refresh</button>"; 

print "</body></html>";   
?>
```

**Reply by Ole-png** [reply]
@ffchung where eg. how do i use it?

**Reply by ffchung** [reply]
hmm... it need put on web server that support php. I just put it on my nas's web server.

**Comment by rekcodocker** [comment]
Question to Original FR Opener.

There is a setting which counts how many faces Immich needs to consider someone worthy of grouping.  Default is 3 but a little higher gets rid of a lot of random bystanders.

Nevertheless, this would still be a welcome feature.   I have similar face recognition on my Synology nas, and this "Yes you have recognized this person in a few photos but he is just a bystander" is really really useful.

**Reply by AllSparkII** [reply]
I know about this setting, but there are two main problems that for my usecase:

1. Having many burst shots of sport events in my gallery, it won't remove those faces without using a pretty high threshold (too high for many of the people I want detected)
2. I use face detected to find the most possible pictures of every person. It is set to a low threshold (2 on my last run) and sorted out later. This way I am able to find ~93% of the pictures I manually tagged with a person before (my Benchmark). With a setting like 5 it falls off to somewhere around 76% if I remember correctly.

Could I make my life easier this way? For sure! But it's not my way of thinking. I would even use a "people-detection" mode instead of face-detection to bring the benchmark value close to 100%. It's just my obscure way of thinking (and the idea for a useful feature)!

**Comment by CyberCreator** [comment]
I disagree with the idea that strangers in photos should be permanently deleted. Instead, I propose that faces of unknown individuals be moved into a separate folder rather than being removed altogether. Over time, a stranger may become an acquaintance. I’ve experienced this countless times—I often find shared photos featuring new people I’ve recently met, yet they originally appeared in my photos taken at festivals, work, university, conferences, and so on. By keeping these faces in a separate folder, I can later review my archive and recognize someone new in old photos from when we were not yet acquainted. This is an incredible opportunity that should not be lost.

**Reply by ffchung** [reply]
For me there will have 3 level, 1. named, 2. unknown that hidden, 3. unknown that can delete. As I have over 600k photos with over 12k persons found. If I just hidden them, and I want to find someone in hidden later, I will need to loop all person that I don't need. And it will be affect the server performance.

**Reply by CyberCreator** [reply]
> For me there will have 3 level, 1. named, 2. unknown that hidden, 3. unknown that can delete. As I have over 600k photos with over 12k persons found. If I just hidden them, and I want to find someone in hidden later, I will need to loop all person that I don't need. And it will be affect the server performance.

I will give my scenario, which is possible:

I meet a person at work. We become friends. We take a joint photo. I load the photo in immich. I tag him in this new joint photo and... suddenly immich informs me that I have old photos with this person, and the person himself was in the archive as "unknown". And I won't have to go through all the old photos. :)

**Reply by AllSparkII** [reply]
The usecase of @CyberCreator wouldn´t need any implementation/change as you could just use the hidden folder, or am I missunderstanding you?

**Reply by ffchung** [reply]
@CyberCreator Everyone's case is different, and I believe that adding a delete face feature will not affect your scenario. I'm not suggesting to replace the hidden feature with delete. The hidden feature will still be available

**Reply by CyberCreator** [reply]
> @CyberCreator Everyone's case is different, and I believe that adding a delete face feature will not affect your scenario. I'm not suggesting to replace the hidden feature with delete. The hidden feature will still be available

I believe that if a feature for targeted deletion of unwanted faces is planned, there should also be an option for targeted addition of faces without having to reprocess the entire photo collection. Additionally, there are cases where face detection fails on certain images, so it would be very useful to allow manual selection and addition of a face to the database.

**Reply by CyberCreator** [reply]
> The usecase of @CyberCreator wouldn´t need any implementation/change as you could just use the hidden folder, or am I missunderstanding you?

The face deletion feature should not replace the hide face function. Currently, I can hide a person who already has a name, and they end up in the general "hidden faces" folder alongside numerous unnamed faces. These functions should be separated, as there have been instances where a friend's face was accidentally hidden with a single click, requiring me to search through the general list to restore it.
Why is there no confirmation prompt when clicking the 'hide person' button?

**Comment by Dat-Ngu** [comment]
Yeah, I feel like this feature set (face recognition) needs an update for better fine-tuning. There is a lot of face detection, but all of it is poor quality, as in random strangers or random objects, which you have to select one by one to delete or remove manually/etc. 

There needs to exist more straightforward methods to clean up like:
- [ ] Ability to Delete a face/person entirely with a button (prob on the web).
- [ ] Ability to separate existing faces/people from strangers and group them in a face/people group (This one properly can be fixed with something or a model for better face detection).
- [ ] A way to not reset everything face recognition but still allow for a rerun of non-named faces/people.

**Reply by AllSparkII** [reply]
I've even thought of a bigger rework, but there is basically no way (known to me) how to do it.

Face Recognition is cool, but "Human Recognition" is cooler. Sometimes peoples faces are hidden, even though they are part of pictures. With the faces tool, you could tag them manually, but (afaik) it would worsen the facial grouping, as you add something to it that isn't a face. And its way, way worse in actually grouping people (Own AI model, trained aside from immich; can't get around sensititvity for clothing f.e.). But this would be the ""best"" (in terms of hits) way to detect humans/animals ^^

**Reply by mertalev** [reply]
> With the faces tool, you could tag them manually, but (afaik) it would worsen the facial grouping, as you add something to it that isn't a face

Manually added faces are ignored during facial recognition.

**Comment by scrapix** [comment]
To process all people ids in a semi-automated manner 

1. extract all needed IDs from here:
'https://immich.domain/api/people'

2.  create a people1.txt file containing
```people1.txt

delete from asset_faces where "personId" = 'd9b47f35-63d4-402d-91f4-46aa6945abdc';
delete from asset_faces where "personId" = '4317fd33-7fa4-47e9-82da-a958e2121403';
delete from asset_faces where "personId" = '905f3138-e33c-4ca5-8868-fa595860ae70';
etc.
```
3. create a people2.txt file containing
```people2.txt
 delete from person where "id" = 'd9b47f35-63d4-402d-91f4-46aa6945abdc';
 delete from person where "id" = '4317fd33-7fa4-47e9-82da-a958e2121403';
 delete from person where "id" = '905f3138-e33c-4ca5-8868-fa595860ae70';
etc.
```
4. execute following commands:
```sh
docker exec -i immich_database psql -U postgres -d immich -v ON_ERROR_STOP=1 -f - < people1.txt
docker exec -i immich_database psql -U postgres -d immich -v ON_ERROR_STOP=1 -f - < people2.txt
```

**Comment by scrapix** [comment]
Faces / people which were imported via EXIF can't be deleted as for today. So you have to go the extra-mile

**Comment by ireun** [comment]
Isn't this feature already implemented? > https://immich.app/docs/api/delete-people
https://github.com/immich-app/immich/pull/19511

**Reply by scrapix** [reply]
oh yes, you might use the api instead of direct database queries. anyway you have to go the extra mile manually by extracting the not-deleted people ids at least.

> Faces / people which were imported via EXIF can't be deleted as for today. So you have to go the extra-mile

**Comment by urbanrider** [comment]
I really need this to help clean up my people page. I have a handful of “fuzzy face people” that are either background faces or low res or partially covered faces (ski goggles have cause lots of issues) and hundreds of different people get added into the “fuzzy faces” even if I untag all these people these fuzzy people keep getting tagged in photos

**Comment by franzqat** [comment]
I would like to ask for some update on this. I have sorted my faces and hidden manually over 1k faces.

Some features that would be nice are:
- Multi select by dragging or click more faces at once.
- Merge more than 5 faces into the same one
- Delete faces that are not wanted (random strangers) - would be good via multi select
